### ISSUE TITLE
Improving statistical approximations to 18 decimals of precision

### ISSUE THREAD

**Issue by Alexangelj** [primitivefinance/solstat#25]
# Description
- todo @Autoparallel could you explain how we are doing this?
---
Yes. In order to do this we are producing a rational function approximation for the CDF and its inverse. Specifically, we generate  higher order approximations than we have now and we also break the domains of definition into smaller subdomains and build approximations on each subdomain and extend to a piecewise solution.

At a lower level, we are essentially using Remez algorithm for minimizing the maximum absolute error for an approximant. We use mathematical software that allows for arbitrary precision in computing values of the target functions and we truncate our results to 18 decimals accuracy.

**Comment by Alexangelj** [comment]
We investigated the better approximations and even achieved 18 decimals of precision, however, the error on the inverse functions grew. This is because the inverse functions are not inverse approximations to our precise approximations...

Overall, it's possible to not rely on the inverse functions in Primitive's other protocols, so we decided to keep the approximations with lower precision. We can still use the inverse functions on-chain to estimate trade outputs, which is really helpful for testing. Optimal trades should be computed off-chain, but the on-chain swaps can happen at a slight loss of about 12 decimal places. Inverse approximations are precise to about 6 decimal places.
